src/Web/Internal/FormUrlEncoded.hs

```diff
@@ -0,0 +1,499 @@
+import           Data.Word
+
+import           GHC.Exts                   (IsList (..))
+import           GHC.Generics
+import           Network.URI                (escapeURIString, isUnreserved,
+                                             unEscapeString)
+
+import Web.Internal.HttpApiData
+
+-- $setup
+-- >>> :set -XDeriveGeneric
+-- >>> :set -XOverloadedLists
+-- >>> :set -XOverloadedStrings
+-- >>> :set -XFlexibleContexts
+-- >>> :set -XScopedTypeVariables
+-- >>> :set -XTypeFamilies
+-- >>> import Data.Either (isLeft)
+-- >>> import Data.List (sort)
+-- >>> data Person = Person { name :: String, age :: Int } deriving (Show, Generic)
+-- >>> instance ToForm Person
+-- >>> instance FromForm Person
+
+-- | Typeclass for types that can be used as keys in a 'Form'-like container (like 'Map').
+class ToFormKey k where
+  -- | Render a key for a 'Form'.
+  toFormKey :: k -> Text
+
+instance ToFormKey ()       where toFormKey = toQueryParam
+instance ToFormKey Char     where toFormKey = toQueryParam
+
+instance ToFormKey Bool     where toFormKey = toQueryParam
+instance ToFormKey Ordering where toFormKey = toQueryParam
+
+instance ToFormKey Double   where toFormKey = toQueryParam
+instance ToFormKey Float    where toFormKey = toQueryParam
+instance ToFormKey Int      where toFormKey = toQueryParam
+instance ToFormKey Int8     where toFormKey = toQueryParam
+instance ToFormKey Int16    where toFormKey = toQueryParam
+instance ToFormKey Int32    where toFormKey = toQueryParam
+instance ToFormKey Int64    where toFormKey = toQueryParam
+instance ToFormKey Integer  where toFormKey = toQueryParam
+instance ToFormKey Word     where toFormKey = toQueryParam
+instance ToFormKey Word8    where toFormKey = toQueryParam
+instance ToFormKey Word16   where toFormKey = toQueryParam
+instance ToFormKey Word32   where toFormKey = toQueryParam
+instance ToFormKey Word64   where toFormKey = toQueryParam
+
+instance ToFormKey Day              where toFormKey = toQueryParam
+instance ToFormKey LocalTime        where toFormKey = toQueryParam
+instance ToFormKey ZonedTime        where toFormKey = toQueryParam
+instance ToFormKey UTCTime          where toFormKey = toQueryParam
+instance ToFormKey NominalDiffTime  where toFormKey = toQueryParam
+
+instance ToFormKey String     where toFormKey = toQueryParam
+instance ToFormKey Text       where toFormKey = toQueryParam
+instance ToFormKey Lazy.Text  where toFormKey = toQueryParam
+
+instance ToFormKey All where toFormKey = toQueryParam
+instance ToFormKey Any where toFormKey = toQueryParam
+
+instance ToFormKey a => ToFormKey (Dual a)    where toFormKey = toFormKey . getDual
+instance ToFormKey a => ToFormKey (Sum a)     where toFormKey = toFormKey . getSum
+instance ToFormKey a => ToFormKey (Product a) where toFormKey = toFormKey . getProduct
+
+-- | Typeclass for types that can be parsed from keys of a 'Form'. This is the reverse of 'ToFormKey'.
+class FromFormKey k where
+  -- | Parse a key of a 'Form'.
```

Comment:
I was looking at `name` in [17.3](https://www.w3.org/TR/html401/interact/forms.html#h-17.3), but that's not right. `id` says it's case sensitive. So I think it's right as it stands.